<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="rr_robotic_arm">


    <xacro:include filename="$(find rr_robotic_arm)/urdf/robot_gazebo.xacro"/> 
    <xacro:include filename="$(find rr_robotic_arm)/urdf/robot_ros2_control.xacro"/> 


    <xacro:macro name="default_inertial" params="mass">
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
    </xacro:macro>


    <material name="white">
        <color rgba="1.0 1.0 1.0 1.0"/>
    </material>
    
    <material name="green">
        <color rgba="0.0 0.6 0.0 1.0"/>
    </material>


    <link name="base_link">
        <visual>
            <geometry>
                <cylinder radius="0.15" length="0.02"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.15" length="0.02"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="link1">
        <visual>
            <geometry>
                <cylinder radius="0.1" length="0.1"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.05" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.1"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.05" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="link2">
        <visual>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.15" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.15" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="link3">
        <visual>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.15" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.15" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="wrist">
        <visual>
            <geometry>
                <box size="0.12 0.08 0.07"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.035" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.12 0.08 0.07"/>
            </geometry>
            <material name="green"/>
            <origin xyz="0.0 0.0 0.035" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="finger_1">
        <visual>
            <geometry>
                <box size="0.02 0.02 0.07"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.025" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.02 0.02 0.07"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.025" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>

    <link name="finger_2">
        <visual>
            <geometry>
                <box size="0.02 0.02 0.07"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.025" rpy="0.0 0.0 0.0"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.02 0.02 0.07"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0.0 0.0 0.025" rpy="0.0 0.0 0.0"/>
        </collision>
        <xacro:default_inertial mass="1"/>
    </link>


    <joint name="joint1" type="revolute">
        <origin xyz="0.0 0.0 0.018" rpy="0.0 0.0 0.0"/>
        <parent link="base_link"/>
        <child link="link1"/>
        <axis xyz="0.0 0.0 1.0"/>
        <limit lower="-1.57" upper="1.57" effort="100.0" velocity="1"/>
    </joint>

    <joint name="joint2" type="fixed">
        <origin xyz="0.0 0.0 0.08" rpy="0.0 0.0 0.0"/>
        <parent link="link1"/>
        <child link="link2"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="joint3" type="revolute">
        <origin xyz="0.0 0.0 0.28" rpy="0.0 0.0 0.0"/>
        <parent link="link2"/>
        <child link="link3"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-1.57" upper="1.57" effort="100.0" velocity="1"/>
    </joint>

    <joint name="joint4" type="fixed">
        <origin xyz="0.0 0.0 0.3" rpy="0.0 0.0 0.0"/>
        <parent link="link3"/>
        <child link="wrist"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="finger_joint1" type="prismatic">
        <origin xyz="-0.04 0.0 0.07" rpy="0.0 0.0 0.0"/>
        <parent link="wrist"/>
        <child link="finger_1"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="0.0" upper="0.03" effort="100.0" velocity="1"/>
    </joint>

    <joint name="finger_joint2" type="prismatic">
        <origin xyz="0.04 0.0 0.07" rpy="0.0 0.0 0.0"/>
        <parent link="wrist"/>
        <child link="finger_2"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-0.03" upper="0.0" effort="100.0" velocity="1"/>
    </joint>

    <gazebo reference="base_link">
        <material>Gazebo/Green</material>
    </gazebo>

    <gazebo reference="link1">
        <material>Gazebo/Green</material>
    </gazebo>

    <gazebo reference="link2">
        <material>Gazebo/White</material>
    </gazebo>

    <gazebo reference="link3">
        <material>Gazebo/White</material>
    </gazebo>

    <gazebo reference="wrist">
        <material>Gazebo/Green</material>
    </gazebo>

    <gazebo reference="finger_1">
        <material>Gazebo/White</material>
    </gazebo>

    <gazebo reference="finger_2">
        <material>Gazebo/White</material>
    </gazebo>

    <!-- Transmission for joint1 -->
    <transmission name="joint1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="joint1_motor">
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    </transmission>

    <!-- Transmission for joint3 -->
    <transmission name="joint3_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="joint3_motor">
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    </transmission>

    <!-- Transmission for finger_joint1 -->
    <transmission name="finger_joint1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="finger_joint1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="finger_joint1_motor">
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    </transmission>

    <!-- Transmission for finger_joint2 -->
    <transmission name="finger_joint2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="finger_joint2">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="finger_joint2_motor">
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    </transmission>

</robot>